### ISSUE TITLE
Azure Keyvault secret reference not resolving within Azure Function connection string

### ISSUE THREAD

**Issue by adetayorabiu** [Azure/Azure-Functions#1291]
Use this repo to file issues in documentation or the Functions Visual Studio tooling. 

If you need technical help, you'll get the fastest response on the forums (https://social.msdn.microsoft.com/Forums/azure/en-US/home?forum=AzureFunctions) or Stack Overflow: http://stackoverflow.com/questions/tagged/azure-functions. The entire engineering team monitors those forums.

- For documentation issues, please provide exactly what is wrong/missing and why it's important. 
- For issues with the Functions Visual Studio tooling, provide detailed steps to reproduce, including details about your environment (OS, VS version, etc)
- For runtime issues (e.g. "my function doesn't work as intended"), use https://github.com/Azure/azure-webjobs-sdk-script
- For issues with the Azure Functions CLI, use https://github.com/Azure/azure-webjobs-sdk-script and add a [CLI] prefix.
- For portal issues, use https://github.com/projectkudu/AzureFunctionsPortal
- For template issues, use https://github.com/azure/azure-webjobs-sdk-templates

If you're not sure where to file your issue, feel free to file it here. Please provide detailed steps to reproduce, including code if applicable.

**Comment by Ohmniox** [comment]
Same issue here. KeyVault secrets are not getting resolved.

**Comment by adetayorabiu** [comment]
@Ohmniox something I figured out is the key vault secret reference doesn't get resolved when used in Connection string configuration and retrieved as IConfiguration.GetConnectionString("conn").

All examples I found has it in AppSettings and retrieved with Environment.GetEnvironmentVariable and that sure works.

 It will be nice to have this limitation documented so we don't go about pulling our hair when secret don't get resolved in connection strings

**Comment by kummerer94** [comment]
This is still happening to me for any Function App on Linux that I try to run (based on Python). I already tried restarting and switched managed identity off and back on.

Is there an update on this issue?

**Comment by SidBarker** [comment]
I have the same issue, Azure Function is not resolving reference to Secrets in KV. 
Do you have any updates about this one? 

I'am using Linux and Python based functions, i did tests on HttpTriggered Functions and EventGridTriggers.

Related discussion on stackoverflow:
https://stackoverflow.com/questions/58384094/access-azure-key-vault-secret-in-azure-function

**Comment by SidBarker** [comment]
After hours of tests I found proper way to resolve this one.

In case of problems with Key Vault Reference make sure that App Function used for Azure Function is based on proper Hosting Plan.

Functions on 'Consumption Plan' are unaable to use Key Vault Reference. Same code on 'App Service Plan' works correctly.

**Comment by jeffhollan** [comment]
To clarify its only Linux Consumption this isn’t yet supported in.

**Comment by kummerer94** [comment]
Thanks, @jeffhollan, for clarifying this. I might have overlooked in the docs. Is there a roadmap or any kind of indication when we can expect this feature? A hosting plan at the moment defeats the purpose of using Azure Functions at least for us.

**Comment by SidBarker** [comment]
@jeffhollan it would be great to find this info in documentation regarding Key Vault Reference :)

**Comment by jeffhollan** [comment]
That purple note at the top of the docs:

https://docs.microsoft.com/en-us/azure/app-service/app-service-key-vault-references

Is on roadmap - I don't have an ETA though.  Totally understand the frustration and is a very useful feature.  For now you'd have to connect and resolve to key vault in code which *works* in the interim but understand it's not ideal

**Comment by kummerer94** [comment]
It's just that you get used to all the awesome features of key vault... but thanks a lot for the information!

**Comment by horeich** [comment]
I have similar issues here: Yesterday I set up a Key Vault and I'm now desperately trying to retrieve secrets from my (currently locally running) .Net Core App without any luck so far. What I've done:
* Setup AAD and an App Registration
* Added a client secret to the registration
* Created new key vault 
The connection to the Key Vault is established withouht problems but the `GetSecretAsync`-method won't resolve it.

Even an old version of my code that used to work with a different subscription stopped getting secrets from my new onmircosoft account. Must be some different config?.

**Comment by dariopad** [comment]
This is usually happening when you change app service plan and the system MI may have changed the GUID. Two ways to resolve: 
1. Just remove the system MI from the KV Access policies and add it back again. (GET is enough)
2. Recreate the System MI from scratch from the App Service/Settings/Identity. Switch off the status and back to on to recreate the GUID. Then, go to AKV and add the access policies for the new GUID. Remove the old one that won't be recognized anymore.
Hope it helps!

**Comment by ClientIco** [comment]
I really think it is unfair to prioritize Linux lower than Windows machines. Honestly, not even having an ETA ? Surely you jest ? Please make this a priority, these gaps cause people to loose days of productive work! @jeffhollan couldn't you atleast speak with the team and hear whether this will be in a years time ? or 6 months ? Also maybe refer us to the roadmap, maybe we can upvote the need for this. After all we are talking about our applications security as well as IP we want protected.

**Comment by Agazoth** [comment]
I have just experienced the exact same issue (not on Linux, though).

I have a function running on Y1, that cannot resolve the @Microsoft.KeyVaul. This is working for me on multiple instances. This time I deployed the environment with a TerraForm template as opposed to the ARM template, VS Code and portal deployments I have done earlier.

I got that annoying red cross on all my secret references. After a few hours of reading documentation and issue reports, I decided to compare it to one of the working setups bit by bit.

After a while I came to the network on the key vault. Apparently I had decided to up the security on this to only allow access to the vault from my current location (uploading keys with PowerShell Secret Management) and AzureServices.

This ought to be all good, as my understanding of the AzureServices bypass switch is, that it allows Azure Services to access the vault in case of network restrictions and default action set to Deny.

```
 (Get-AzKeyVault -VaultName "MyKeyVault").NetworkAcls

DefaultAction                 : Deny
Bypass                        : AzureServices
IpAddressRanges               : {85.184.122.12/32}
IpAddressRangesText           : 85.184.122.12/32
VirtualNetworkResourceIds     :
VirtualNetworkResourceIdsText :
```
Just to make sure, that this was nat ohat was causing my issue, I flipped the switch to allow All networks.

```
DefaultAction                 : Allow
Bypass                        : AzureServices
IpAddressRanges               : {85.184.122.12/32}
IpAddressRangesText           : 85.184.122.12/32
VirtualNetworkResourceIds     :
VirtualNetworkResourceIdsText :
```
And suddenly my function was able to resolve the keys.

The question now is, why does my deployment not allow the function to access the key vault, even though bypass for Azure Services is enabled? According to the documentation [here]( https://docs.microsoft.com/en-us/azure/key-vault/key-vault-overview-vnet-service-endpoints#trusted-services) Azure App Service is is a trusted Azure Service.

I'm not sure if these observations are relevant i this issue, but under all circumstances I would like to know, why I experience this behavior.

**Comment by DarinMacRae** [comment]
I ran into this issue today. In our case we use an Azure AD Group to include our various function apps that need access the our Key Vault. I set this all up as it should work and it does work if you access the key vault via code but using this configuration feature it didn't.

On a hunch I created a new KV access policy that referenced the function app identity directly and then it worked. Is there some reason that AAD groups aren't supported? @jeffhollan 

The reason that using a group is important to us is that early on we created KV access policies for several dozen function apps and we hit an internal limit on access policies. We were told by MS support to use an AAD group as a work-around. This has been working for us for years to allow accessing our KV from code but today I wanted to try this neat configuration feature and was disappointed.

**Comment by Cdoherty95** [comment]
I am having the same issue as described by @DarinMacRae. Assigning the managed identity direct permission in the key vault's access policy works, however, using an AD group does not work. The result description I found in the diagnostics setting log reads `"The user, group or application 'appid=00******-****-****-****-**********00;oid=ce******-****-****-****-**********28;iss=https://sts.windows.net/10******-****-****-****-**********95/' does not have secrets get permission on key vault 'keyvaultname;location=centralus'. For help resolving this issue, please see https://go.microsoft.com/fwlink/?linkid=2125287"`

**Comment by mattchenderson** [comment]
- Networking is something we are still working on. It's a problem with the outbound resolution path and how the traffic there gets routed, basically. But it's something we're actively trying to address
- Group resolution should work. Mechanically, this is the same as use of any other managed identity or principal. Group checks from the access policy side should perform a transitive isMemberOf check. It sounds like that is the case for a manual interaction, but it's not working for the references feature, correct? My quick theory is that this would be due to caching of the token without the group membership claims from the perspective of the referencing component, but that would go away quickly enough and shouldn't present an ongoing issue. We'll look into that.

**Comment by erwelch** [comment]
@mattchenderson - with regards to the networking issue - does this mean we can't use key vault references for a key vault with a private endpoint? I created a private endpoint for my key vault and the Azure Function is integrated into the VNET. It's able to resolve other private endpoints within the same VNET, but it's unable to resolve the key vault reference once I added the private endpoint.

**Comment by mattchenderson** [comment]
Correct. The problem is that the resolution is occurring before the join to the VNET that your code is able to leverage.

**Comment by adiun** [comment]
@mattchenderson thanks for the update. Do you have an ETA on when the networking problem would be addressed? My team has a VNet infra in place and we'd like to continue using KeyVault references.

**Comment by mattchenderson** [comment]
Unfortunately, I do not have an ETA I can share at this time.

**Comment by oliverwhk** [comment]
this issue has been around for a year, is it just not high priority enough to have it resolved?

**Comment by jeffhollan** [comment]
Was a twitter exchange on this today. It’s high priority and in flight, but often these networking features require many dependent pieces to all fall into place and then a larger upgrade of our fleet so do take a fair bit more time than a bug fix on the runtime for example. But in coming months should allow you to resolve key vaults using this pattern behind networking constraints. https://twitter.com/ccompy/status/1309883379228131329?s=21

**Comment by thdotnet** [comment]
Thanks for the update @jeffhollan . You point the purple bar in the official doc (https://docs.microsoft.com/en-us/azure/app-service/app-service-key-vault-references) however, it doesn't contain a workaround. What should we do while this feature is not released? Allow network traffic as @Agazoth  did? Use code to retrieve the secrets rather than Key Vault reference?

**Comment by jeffhollan** [comment]
@thdotnet yes either of those would work. Trigger secrets would be the exception for the piece that you wouldn’t have a viable workaround for - but rest should - though obviously not ideal

**Comment by massimo1993** [comment]
@thdotnet I think for the network issue you just need to include the list of IP addresses for the function app into the Key Vault firewall. I tested it by adding a whitelist of the outbound IP addresses generated from the function app.

https://docs.microsoft.com/en-us/azure/azure-functions/ip-addresses

I use Terraform to manage my infrastructure so these articles might be of help.

https://www.terraform.io/docs/providers/azurerm/r/function_app.html#outbound_ip_addresses
https://www.terraform.io/docs/providers/azurerm/d/function_app.html#outbound_ip_addresses

**Comment by majorvin** [comment]
@jeffhollan im getting the issue as well but i can confirm it works in linux. The workaround is to recreate the application setting that is referenced a keyvault. Although, this is not really a workaround if we are using pipeline to deploy the function app.

**Comment by MarkAtAgilliance** [comment]
Grrr I wish I had read this article instead of the docs before changing 15 services to use KV references instead of code.  We have app services plans and use groups for permissions and vnets.  Failed every test, now rolling back, re-deploying and re-testing.  What a waste of time.  IMHO KeyVault related access schemes should be first class citizens of security.  A work around that says "open up the networking" I don't think is helpful.

**Comment by dlandi** [comment]
This is frustrating. Can someone post the exact conditions in which this *does* work?

Is there a happy path?

**Comment by balag0** [comment]
@dlandi Could you share your appname. 
https://github.com/Azure/azure-functions-host/wiki/Sharing-Your-Function-App-name-privately

**Comment by schaijkw** [comment]
> Was a twitter exchange on this today. It’s high priority and in flight, but often these networking features require many dependent pieces to all fall into place and then a larger upgrade of our fleet so do take a fair bit more time than a bug fix on the runtime for example. But in coming months should allow you to resolve key vaults using this pattern behind networking constraints. https://twitter.com/ccompy/status/1309883379228131329?s=21

Hi @jeffhollan , a few months have past. Is there an ETA? I have to store my SendGrid ApiKey for a FunctionApp and am willing to use AzureWebJobsSendGridApiKey to pass the ApiKey for the SendGrid output binding. Due to sensitive data we process we cannot open up our network. Is there another way to workaround this issue so I can use a secret in the FunctionApp output binding?

**Comment by Nlouis00100** [comment]
Hi, ETA on this feature. Very high priority for us. Thank you.

**Comment by jeffhollan** [comment]
Updating here:

- key vault reference should now work in Linux consumption
- Key vault references no longer need to be pinned to a version, but can point to a key and will be updated on new versions (not instant, check the docs)
- Key vault references won’t work when the key vault is behind a private network
- 

Support for private network references is rolling out globally in the coming weeks. Keep an eye out on azure updates for that.

**Comment by Nlouis00100** [comment]
Thank you for the feedback.
Are you saying the keyvault inside a vnet will soon support reference call
(app settings) from webapp/funtion? I haven’t seen any update on this
limitation anywhere (other than it does not work).
Nicolas

**Comment by jeffhollan** [comment]
Correct.  The `>IMPORTANT` callout on the doc calls out this current limitation.  No concrete ETA but work / deployments are in flight to remove this restriction https://docs.microsoft.com/en-us/azure/app-service/app-service-key-vault-references

**Comment by leonK-DI** [comment]
Hello 

I stiill cannot reference secrets in my keyvault. Is this reference feature as well rolloed out to Europe West?

**Comment by balag0** [comment]
Yes. It is rolled out in West Europe. Could you share your app name if you are having issues with this - https://github.com/Azure/azure-functions-host/wiki/Sharing-Your-Function-App-name-privately

**Comment by leonK-DI** [comment]
> Yes. It is rolled out in West Europe. Could you share your app name if you are having issues with this - https://github.com/Azure/azure-functions-host/wiki/Sharing-Your-Function-App-name-privately

Here we go: 2021-02-22T04:08:28Z   [Information] (Id=f4c91f58-cb63-406c-95f5-cfe76188ea49)

**Comment by balag0** [comment]
Looks like there is some issue with the configured permissions. (I have removed the actual appid and keyvault names in the below exception) - https://docs.microsoft.com/en-us/azure/key-vault/general/assign-access-policy-portal

WebServiceError.Code:Forbidden WebServiceError.Message:The user, group or application 'appid=b0ca..' does not have secrets get permission on key vault '[Keyvaultname];location=westeurope'. For help resolving this issue, please see https://go.microsoft.com/fwlink/?linkid=2125287

**Comment by leonK-DI** [comment]
> Looks like there is some issue with the configured permissions. (I have removed the actual appid and keyvault names in the below exception) - https://docs.microsoft.com/en-us/azure/key-vault/general/assign-access-policy-portal
> 
> WebServiceError.Code:Forbidden WebServiceError.Message:The user, group or application 'appid=b0ca..' does not have secrets get permission on key vault '[Keyvaultname];location=westeurope'. For help resolving this issue, please see https://go.microsoft.com/fwlink/?linkid=2125287

I turned on the system managed identity for my function & an access policy is as well added. But I will try to delete the access policy and add it again. Let's see if this helps...

**Comment by mzhukovs** [comment]
@jeffhollan Apologies still have trouble confirming what is coming out imminently (or should already be working). I have a Key Vault that is behind a vnet in West Europe, and a Windows _Consumption_-based .NET Azure Functions App - I added the Function App's MSI to the Key Vault with Key Reader role with Get/List permissions on secrets etc.  - am still getting this error: 

> **AccessToKeyVaultDenied:** Key Vault reference was not able to be resolved because site was denied access to Key Vault reference's vault.

Is the fix that's coming applicable to this scenario, or is it only addressing some other issue for **non**-consumption-based plans?

Appreciate the support.

**Comment by leonK-DI** [comment]
@mzhukovs you have to add an "Access Policy" to your vault and not ad a role to the managed identity of your app. At least that solved for me the issue

**Comment by mzhukovs** [comment]
> @mzhukovs you have to add an "Access Policy" to your vault and not ad a role to the managed identity of your app. At least that solved for me the issue

Yes I did both trying to get it to work. Access policy for the function app's managed identity AND Azure key vault reader role. Still no luck.

**Comment by onionhammer** [comment]
One thing I would find super useful is being able to reference secrets in parameters directly without a .params.json file; which are super verbose and have to be mapped exactly 1:1 to existing parameters in the bicep file.

```
param mySomething string {
     secure: true // or default to true
     secret: https://my-vault.vault.azure.net/secrets/my-secret/39179d873f354b4ca8f4f91a44d4c997 // Copy-paste diretly from vault in portal
}

**Comment by mzhukovs** [comment]
https://feedback.azure.com/forums/355860-azure-functions/suggestions/38817385-allow-key-vault-references-to-access-secrets-behin

**Comment by konjp** [comment]
@oliverwhk Have you resolved your [issue](https://developercommunity.visualstudio.com/t/azure-pipeline-managed-identity-unable-to-find-the/1187791)? I still have the same thing with `AzureWebJobsStorage ` put it in key vault, and fails when deployed using `AzureFunctionApp@1` with error "_Unable to find the storage account associated with the function app_". No firewall rules, no VNET configured. Strange that there no more people asking for this, because should not this be recommended way to store access key in key vault?

**Comment by oliverwhk** [comment]
@konjp i haven't tried since then 😅

**Comment by konjp** [comment]
I think problem in this case that you @oliverwhk  referenced, and I confirmed that still exists. is between Azure DevOps and Azure Cloud.  And "nobody's" problem is hard to resolve.

Only I can you do for now is not use key vault for _AzureWebJobsStorage_ property.

**Comment by jamyc** [comment]
We're currently facing an issue where one of our Azure Functions doesn't appear to be resolving the KeyVault Reference correctly. Not sure if this is the correct function id, but couldn't find it using the link posted above (perhaps because we're running a dotnet-isolated function?): `Id=3474fdc1-a852-4f65-b4b3-df00e6188f0c`

It's a dotnet-isolated Azure Function which has an `EventHubTrigger`. I get an exception in Application Insights like:
`Error indexing method 'XXX' Illegal connection string parameter name '@Microsoft.KeyVault(SecretUri' (Parameter 'connectionString')`

This appears to be thrown from [this](https://github.com/Azure/azure-event-hubs-dotnet/blob/dev/src/Microsoft.Azure.EventHubs/Primitives/EventHubsConnectionStringBuilder.cs#L337) line in the `EventHubsConnectionStringBuilder.cs` which makes me believe that the KeyVault reference is not properly being resolved.

This is especially interesting since the KeyVault reference is giving us a green checkmark on the Application settings page:
![image](https://user-images.githubusercontent.com/6288420/135648561-f83f08f2-04ef-45dd-8a6d-49b2cf160ee4.png)

@balag0 Could you perhaps have a look at this? Happy to PM/mail you the instance details if the id posted above doesn't provide you with any results.

**Comment by balag0** [comment]
@jamyc Sure, could you please open a new issue with these details and tag me since this is a closed issue.

**Comment by pawel-zolty** [comment]
> * Networking is something we are still working on. It's a problem with the outbound resolution path and how the traffic there gets routed, basically. But it's something we're actively trying to address
> * Group resolution should work. Mechanically, this is the same as use of any other managed identity or principal. Group checks from the access policy side should perform a transitive isMemberOf check. It sounds like that is the case for a manual interaction, but it's not working for the references feature, correct? My quick theory is that this would be due to caching of the token without the group membership claims from the perspective of the referencing component, but that would go away quickly enough and shouldn't present an ongoing issue. We'll look into that.

After 20 hours AAD group still not working for me.
I add functionApp system MSI to security AAD group and give it access policy at key vault. It does not work at all...

**Comment by himanshu-ram-punjabi** [comment]
Still having the same issue like @DarinMacRae, @Cdoherty95 and @pawel-zolty. If the System Assigned Managed Identity of the App Service / Function app is in a Security Group and the Security Group is assiged a Key Vault Administrator RBAC Policy on the Key Vault, the Key Vault References don't work in Environment Variables. Direct assignment of the Identity to the Key Vault seem to be working fine.

**Comment by MarkAtAgilliance** [comment]
You may need to wait if using SGs.  The token on the client might need to refresh if cached.  Restarting the app service host can force it but I have had mixed results and sometimes it just starts working.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by himanshu-ram-punjabi** [comment]
I did restart the Function app and also did a 'Pull Reference Values' but still given me the same error: 
Key Vault reference was not able to be resolved because site was denied access to Key Vault reference's vault.

Let's see. Shall wait for some more time else I'll do a direct assignment removing it from the group.

**Comment by MarkAtAgilliance** [comment]
I would keep the group assignment/membership for now *and* add the direct assignment.  Then later you can remove the direct assignment to see if it works after your token was updated.

**Comment by himanshu-ram-punjabi** [comment]
Thank you so much @MarkAtAgilliance. I check now and I can see it is working with the Group assignment after ~8 hours. Adding the Direct assignment was not required. So, it must be related to the refresh of the Token and caching.

**Comment by himanshu-ram-punjabi** [comment]
Instead of restaring the Web App, re-Deploying the code to the Web App clears the cache instantly and the app starts referencing the Key Vault secrets, when the System Assigned Managed Identity of the app is part of a Security Group that has a Role Assignment on the Key Vault. Sharing as I came across this instance today.

**Comment by himanshu-ram-punjabi** [comment]
I ran into the same problem today, with the Key Vault References not working after another deployment with no changes done to the way the Identity was configured i.e. still part of a Group and the Group has the required permission on the Key Vault. I read an article by Microsoft [here](https://learn.microsoft.com/en-us/entra/identity/managed-identities-azure-resources/managed-identity-best-practice-recommendations#limitation-of-using-managed-identities-for-authorization) and it does talk about the Identity Token and the time it takes for it to refresh (which is nearly 24 hours). So, sticking to granting the System Assigned Identity a Role directly on the Key Vault instead of keeping it in the group for now.
